### ISSUE TITLE
Give access to re-apply share button to consumers 

### ISSUE THREAD

**Issue by anushka-singh** [data-dot-all/dataall#1286]
**Is your idea related to a problem? Please describe.**
Currently, only a dataset owner/approver has access to fix a share by clicking the re-apply share button. 
If a requestor finds that their share is in unhealthy state, they need to reach out to dataall admins or the dataset owners which causes an unnecessary wait cycle when a production system share could be broken. 

**Describe the solution you'd like**
I propose giving access to the requestor team to the re-apply share button. A share would be a joint responsibility of a producer and consumer. Consumers should be able to hit the button and fix their own shares without relying on anyone else to fix it. We additionally would like the producers to get notified via email + UI that their share had broken and was in unhealthy state and fixed by a consumer. This would ensure that the error message from unhealthy status is also captured and sent to producers so they can fix any underlying issues that caused the problems instead of just bandaging them. 

*P.S. Don't attach files. Please, prefer add code snippets directly in the message body.*

**Comment by noah-paige** [comment]
Hi @anushka-singh - for the notification aspect of the above issue, let's use #1420 as the parent issue to improve all types of notifications on data.all

For reapply - it is true only Approvers are able to reapply a share at the moment intentionally so. Because `re-apply` is going through very similar motions to `approve` and potentially would be editing the permissioning on the data resources it felt safest to restrict re-applying only to the Approver persona in the share

That being said I do understand the alternative line of thought that once a share is approved it is a shared responsibility to manage the health of that share. I can not think of current risks of doing so - @anmolsgandhi @dlpzx  do you see any inherit risk here with this approach?

If not this could be a nice enhancement to the current functionality

**Comment by TejasRGitHub** [comment]
# Implementation Ways 

## Proposed way 

Frontend change - Update the ShareView.js file with conditional to show reapply share button 
Backend change - The reapply share item mutation is decorated with the APPROVE_SHARE_OBJECT permission i.e. only groups / teams have the approve share object permission would be able reapply the shares. This permission is not attached to the requestors / consumers.  For this feature enhancement, this permission on the decorator can be switched to GET_SHARE_OBJECT ( just like verify share object function ). Additional check in the `reapply_items_share_object` method can be added to only allow reapplying only those share object items which are in [`Share_Succeeded`, `Share_Failed`, `Share_Approved`] state as only those share items should be reapplied. 

LOF - 2 days

## Alternative option

Same frontend change as the previous
Backend change - 
1. Introduce another permission 'REAPPLY_SHARE_OBJECT', and assign this permission to the decorator of the `reapply_items_share_object` method. 
2. Create an alembic script to backfill this permission into all the requestor and the approver teams to have access to their share for reapplying share. 
3. Also, with this, the create , update share object methods would need an update to include / attach this permission to the requestor and approver teams.

LOF - 4 days

**Comment by petrkalos** [comment]
`GET_SHARE_OBJECT` mentally points to a read only operation (like `verify_share`) so I am not very keen on reusing the same permission. I think it will be clearer to introduce a new permission for these kind of situations. 

Perhaps instead of naming it `REAPPLY_SHARE_OBJECT` give it a broader name (maybe `MODIFY|AMEND|EDIT_SHARE_OBJECT` ) so that we can reuse it in the future for other stuff.

**Comment by dlpzx** [comment]
I agree on having the alternative option with a more specific permission separate from the read-only GET_SHARE_OBJECT. Not strongly opinionated on the name of the new permission, something generic like AMEND might make it easy in the future to add new things
